Fixes mocking inherited class methods

nielsvaneck opened the pull request:
Fixes #33 

Mockito loses track of the original_method. Then when the new mocked method is invoked, then invocation is handled as a class method, the first argument is removed, which is what eventually causes the TypeError that shows up in the failing tests. 

This change finds the original method if a classmethod is brought in through inheritance.  It also addresses a test failure against Python 2.7 that showed up after fixing the initial problem.

nielsvaneck commented:
Passes all tests on all configured Python versions: https://travis-ci.org/kaste/mockito-python/builds/651166706

kaste commented:
Very interesting and thank you.  My first intuition was this must be a regression but I think it is not. I will leave some comments.

kaste commented on mockito/mocking.py:
```diff
@@ -79,7 +79,10 @@ def get_original_method(self, method_name):
+            if not original_method:
+                original_method = getattr(self.spec, method_name, None)
+            return original_method
         except AttributeError:
```
Here the original code a bit dubious.  E.g. why is it okay to return None here? T.i. why do we have the safe `get(method_name, None)`?  

Wouldn't it be better when we unsafely try to get `self.spec.__dict__[method_name]` and then just catch both `except (AttributeError, KeyError):` to do the fallback `getattr(...)`

kaste replied:
Iirc `original_method == None` is a marker that the method will be *added* (and potentially shadow a parent method) and then will affect the handling when unstubbing.

nielsvaneck replied:
Yes, I was thinking the same. 

dict.get actually automatically falls back to None, which is what caused this in the first place. 

I can't see how (`return self.spec.__dict__.get(method_name)`)[https://github.com/kaste/mockito-python/blob/79ae4340635c8298a79b1d6afc8e6b0b5ce26f88/mockito/mocking.py#L82] would ever throw an AttributeError, so if __dict__ doesn't contain what we're looking for, we end up with None.

I also tried always using `getattr(self.spec, method_name, None)`  but that ended up failing a whole bunch of tests ¯\_(ツ)_/¯

kaste commented on tests/classmethods_test.py:
```diff
@@ -22,12 +22,25 @@
+        return bark
+
+
+class Dog(LoudBarker):
```
I don't like it if we change the `Dog` structure and meaning for the old tests.  Esp. if you look below when we explicitly do a old-school `Lion(object)` it looks like inheritance could matter.  So I think it would be better to define two classes just for this purpose.  

To test the wanted behavior, it is also important that we check that we actually don't change or mock the base class `LoudBarker` when mocking the `Barker`. Basically in pseudo-code

```
when(Barker).loud_bark....
assert LoudBarker.loud_bark(...)  # ensure not mocked
```

And that we get the old classes back after unstubbing. That's really important for inheritance mocking because usually we shadow the original method when mocked, and need to *delete* the shadowing classmethod on unstub. In contrast: replacing it with the original method would not be the same because then a mocked base `Barker` would probably not change the behavior of the `LoudBarker` anymore.

nielsvaneck replied:
Agreed.  I'll rework this.

nielsvaneck commented:
I reworked the tests.  https://travis-ci.org/kaste/mockito-python/builds/651216907

lmk what you think.

nielsvaneck commented on tests/classmethods_test.py:
```diff
@@ -105,3 +106,107 @@ def testPreservesClassArgumentAfterUnstub(self):
+    # TODO decent test case please :) without testing irrelevant implementation
+    # details
+    #
+    # TODO This test does not work for inherited @classmethods. but that does
```
That comment is based on the following finding in the Python interpreter: 

```
>>> class Dog():
...     @classmethod
...     def bark(cls):
...         return "woof!"
... 
>>> class Retriever:
...     @classmethod
...     def retrieve(cls, item):
...         return item
... 
>>> 
>>> class TrickDog(Dog, Retriever):
...     pass
... 
>>> isinstance(Dog.__dict__.get("bark"), classmethod)
True
>>> isinstance(TrickDog.__dict__.get("bark"), classmethod)
False
>>> 
```

kaste replied:
Hm, yeah sure, the `dict` of `TrickDog` is basically empty and `None`is not a `classmethod`.

kaste commented on tests/classmethods_test.py:
```diff
@@ -105,3 +106,107 @@ def testPreservesClassArgumentAfterUnstub(self):
+        verify(TrickDog).bark()
+        verify(TrickDog).retrieve("stick")
+
+    def testPreservesClassArgumentAfterUnstub(self):
```
I think this is the same test `testUnstubs` above.

nielsvaneck replied:
you're right.  I'll remove it.

kaste commented:
I liked that simple patch, unfortunately 

```
    def testA(self):
        when(Retriever).retrieve("stick").thenReturn("ball")
        when(TrickDog).retrieve("stick").thenReturn("cat")
        unstub()
        self.assertEqual("stick", TrickDog.retrieve("stick"))
```

fails. 

![image](https://user-images.githubusercontent.com/8558/74683832-694b2f00-51ca-11ea-9b23-ce29788412db.png)

Unless I'm confused after that setup `TrickDog` is corrupted. 😞

nielsvaneck commented:
that is unfortunate indeed!   (from a practical point of view, it's still a good bit better than the behavior it fixes IMO)

TrickDog might be fine.  Does unstub fail to remove the modification from Retriever?  let me dig a little.

kaste commented:
No it replaces `TrickDog.retrieve` with the patched `Retriever.retrieve` because that was the "original" method at that time.  But unstubbing must *delete* the classmethod from `TrickDog` because we basically added it. No?

nielsvaneck commented:
Right!  So if we want to stick with the approach of stubbing at the "top" level, that leaves us with 2 choices: 
1)  when dealing with `when(TrickDog).retrieve...`  we have to actually find `retrieve` on `Retriever` and stub it there, because that's technically where the real method lives

or 

2) somehow remembering that when we're stubbing `retrieve` on `TrickDog` it technically doesn't replace anything, in which case all `unstub()` needs to do  is just remove the stub.

kaste commented:
It's the second one otherwise a `TrickierDog` would see the mock as well.

nielsvaneck commented:
Good point.  There seems to be a a bit of support for stubbing "nonexistent" methods, based on "original_method" being None (as you mentioned earlier)   but that's also what got us into this pickle to start with.. 

Maybe we can use original method while setting up the stub, but then not store it?

kaste commented:
It should be easy to return a tuple `(original_method: str, added: bool)` from `get_original_method` and then use that information in `restore_method`.  (I'm currently hitting a bug with the newest git version I just installed, so I have to restore git functionality first.)

nielsvaneck commented:
I'll give that a try

kaste commented:
That's again a nice patch.  It's also very explicit via the comments that we don't know a lot here and don't model the how-to-undo-the-stubbing rather implicit. I tried to break it without success. 👍 

There is a small change in semantics for the `AttributeError` case because after this patch we always have `was_in_spec=False`. Before we set this only if we returned `None`. Hard to find a test for this, afaik as long as we don't have a `.__dict__` we also can't patch anyway. E.g. 

```
class Point:
	__slots__ = ("x", )
	def double(self): pass
```

Instances of `Point` have no `__dict__` (bc `slots`) but then `double` becomes read-only.

```
----> 1 setattr(p, "double", lambda: "foo")

AttributeError: 'Point' object attribute 'double' is read-only
```

As I said everything looks good. There are some typos in the method doc. I also tested with this

```
    def testUnStubWorksOnClassAndSuperClass(self):
        self.assertEqual("stick", Retriever.retrieve("stick"))
        self.assertEqual("stick", TrickDog.retrieve("stick"))

        when(Retriever).retrieve("stick").thenReturn("ball")
        self.assertEqual("ball", Retriever.retrieve("stick"))
        self.assertEqual("ball", TrickDog.retrieve("stick"))

        when(TrickDog).retrieve("stick").thenReturn("cat")
        self.assertEqual("ball", Retriever.retrieve("stick"))
        self.assertEqual("cat", TrickDog.retrieve("stick"))

        unstub(TrickDog)
        self.assertEqual("ball", Retriever.retrieve("stick"))
        self.assertEqual("ball", TrickDog.retrieve("stick"))

        unstub(Retriever)
        self.assertEqual("stick", Retriever.retrieve("stick"))
        self.assertEqual("stick", TrickDog.retrieve("stick"))

    def testReverseOrderWhenUnstubbing(self):
        when(Retriever).retrieve("stick").thenReturn("ball")
        when(TrickDog).retrieve("stick").thenReturn("cat")

        unstub(Retriever)
        self.assertEqual("stick", Retriever.retrieve("stick"))
        self.assertEqual("cat", TrickDog.retrieve("stick"))

        unstub(TrickDog)
        self.assertEqual("stick", Retriever.retrieve("stick"))
        self.assertEqual("stick", TrickDog.retrieve("stick"))

```

which is rather verbose but maybe better explains the "inheritance" when mocking and partially unstubbing.

nielsvaneck commented:
good morning! thank you for reviewing, i’ll incorporate your suggested tests. 

shall I remove the one that is commented out? (with the comment about inherited class methods not appearing in __dict__)

regarding the typos you mentioned, I read through the code again just now and nothing immediately jumps out at me as a glaring typo. perhaps it’s author goggles, or maybe it’s due to a difference in experience and verbal model of the subject matter. in any case, would you mind putting some comments in the diff on wording you’d like to see fixed? i’m happy to adopt whatever changes you suggest.

nielsvaneck commented:
btw, just thinking about that `AttributeError` again. in what situation would we see that?

kaste commented on mockito/mocking.py:
```diff
@@ -77,7 +77,7 @@ def clear_invocations(self):
         """
         Looks up the original method on the `spec` object and returns it
-        together with an indicatin of whether the the method is found
+        together with an indication of whether the the method is found
```
yeah, it was this plus the double "the". 😄  Minor things.

nielsvaneck replied:
d'oh!  fixed.

nielsvaneck replied:
a clear case of "author-goggles"  ; )

kaste commented:
We probably don't need the commented test.  

Do you know any other ways to get an object without a `__dict__`? Above I posted the only way I know using `slots` but this doesn't mean you can mock such a method. But it provides a clear error message.  Without the `AttributeError` catch we would crash in a way a user can't understand. It would look like a mockito bug. With it we clearly say 'x' is read-only.

nielsvaneck commented:
I removed the commented test. 

Also added a little comment to clarify the AttributeError. That might be helpful for the next person who comes along.. 

I cannot think of a way to make this work with slots.  unless you specify the method name you'd want to stub out in the slots collection, there's no way to overwrite them.

kaste commented:
Good to go, released as 1.2.1 Thanks for making this.

kaste commented:
Everything okay?

nielsvaneck commented:
I updated mockito in my project and it works like a charm. thank you!
